cogeo_mosaic/supermorecado.py

```diff
@@ -0,0 +1,173 @@
+"""Supermercado.burntiles but for other TMS.
+
+This submodule is adapted from mapbox/supermercado project:
+
+The MIT License (MIT)
+
+Copyright (c) 2015 Mapbox
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+
+"""
+
+from typing import Any, Dict, Sequence, Tuple
+
+import attr
+import morecantile
+import numpy
+from affine import Affine
+from rasterio import features
+
+WEB_MERCATOR_TMS = morecantile.tms.get("WebMercatorQuad")
+
+
+def _feature_extrema(geometry: Dict) -> Tuple[float, float, float, float]:
+    """Get bounds extrema for a feature."""
+    if geometry["type"] == "Polygon":
+        x, y = zip(*[c for part in geometry["coordinates"] for c in part])
+
+    elif geometry["type"] == "MultiPolygon":
+        x, y = zip(
+            *[c for poly in geometry["coordinates"] for part in poly for c in part]
+        )
+
+    elif geometry["type"] == "LineString":
+        x, y = zip(*[c for c in geometry["coordinates"]])
+
+    elif geometry["type"] == "Point":
+        x, y = geometry["coordinates"]
+        return x, y, x, y
+
+    return min(x), min(y), max(x), max(y)
+
+
+def find_extrema(
+    features: Sequence[Dict[Any, Any]]
+) -> Tuple[float, float, float, float]:
+    """Get bounds extrema for a set of features."""
+    epsilon = 1.0e-10
+    min_x, min_y, max_x, max_y = zip(
+        *[_feature_extrema(f["geometry"]) for f in features]
+    )
+
+    return (
+        min(min_x) + epsilon,
+        max(min(min_y) + epsilon, -85.0511287798066),
+        max(max_x) - epsilon,
+        min(max(max_y) - epsilon, 85.0511287798066),
+    )
+
+
+@attr.s
+class burnTiles:
+    """Burntiles."""
+
+    tms: morecantile.TileMatrixSet = attr.ib(default=WEB_MERCATOR_TMS)
+
+    @tms.validator
+    def _check_for_quadtree_support(self, attribute, value: morecantile.TileMatrixSet):
+        if not value._is_quadtree:
+            raise ValueError(f"{value.identifier} TMS does not support quadtree.")
```

Comment:
supporting more than just quadtree structures would be my hope here as equirectangular codes for planetcantile current use 2,1 matrix scales. Is there a limitation of the burn code I'm not seeing here that excludes other scales?

Reply:
you're right quadkey support is only needed for mosaic, not for burntiles 👍